### ISSUE TITLE
Unsound Send and Sync impls when the borrowed type is not

### ISSUE THREAD

**Issue by dtolnay** [Voultapher/self_cell#4]
Filing to track the bug reported by @jDomantas in https://www.reddit.com/r/rust/comments/m42fjx/safetouse_procmacrofree_selfreferential_structs/gqt3i35/, which can result in UB in safe code.

```rust
use std::rc::Rc;
use once_self_cell::sync_once_self_cell;

struct Owner;

struct Borrowed<'a> {
    owner: &'a Owner,
    evil: Evil,
}

impl<'a> From<&'a Owner> for Borrowed<'a> {
    fn from(owner: &'a Owner) -> Borrowed<'a> {
        Borrowed {
            owner,
            evil: Evil::default(),
        }
    }
}

thread_local! {
    static STR: Rc<str> = "hello".into();
}

#[derive(Clone)]
struct Evil {
    rc: Rc<str>,
}

impl Drop for Evil {
    fn drop(&mut self) {
        let ptr = self.rc.as_ptr();
        let thread_id = std::thread::current().id();
        println!("drop rc at {:p} in thread {:?}", ptr, thread_id);
    }
}

impl Default for Evil {
    fn default() -> Evil {
        Evil {
            rc: STR.with(|x| x.clone()),
        }
    }
}

sync_once_self_cell!(
    SelfRef,
    Owner,
    Borrowed<'_>,
);

fn main() {
    let cell = SelfRef::new(Owner);
    let _evil1 = cell.get_or_init_dependent().evil.clone();
    std::thread::spawn(move || {
        let _evil2 = cell.get_or_init_dependent().evil.clone();
    });
}
```

```console
drop rc at 0x21b53dbd0d0 in thread ThreadId(1)
drop rc at 0x21b53dbd0d0 in thread ThreadId(2)
drop rc at 0x21b53dbd0d0 in thread ThreadId(2)
```

**Comment by Voultapher** [comment]
See this comment chain for more background and discussion of different approaches https://www.reddit.com/r/rust/comments/m42fjx/safetouse_procmacrofree_selfreferential_structs/gqt3i35/

I'll fix it once I find the time.

**Comment by Voultapher** [comment]
This gets addressed by #7, the relevant place are:

- https://github.com/Voultapher/once_self_cell/pull/7/files#diff-9ab425d0ba4c43d4de2ad95d93c99900e0bde023ea43a86cfae39888fc20266aR135
- https://github.com/Voultapher/once_self_cell/pull/7/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R355
- https://github.com/Voultapher/once_self_cell/pull/7/files#diff-1778ecc1b76a790dccfe3d31d658c07d2464493912a0e873cce3cdf1a361f8d7R255
